pkg/agent/multicast/mcast_controller.go

```diff
@@ -0,0 +1,366 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+)
+
+type eventType uint8
+
+const (
+	groupJoin eventType = iota
+	groupLeave
+
+	// Use IGMPv3 for IGMP query message because it is the main version for IGMP.
+	// TODO: support sending IGMP query with both v2 and v3 messages
+	queryVersion      = 3
+	podInterfaceIndex = "podInterface"
+
+	// How long to wait before retrying the processing of a multicast group change.
+	minRetryDelay = 5 * time.Second
+	maxRetryDelay = 300 * time.Second
+)
+
+var (
+	workerCount uint8 = 2
+)
+
+type mcastGroupEvent struct {
+	group net.IP
+	eType eventType
+	time  time.Time
+	iface *interfacestore.InterfaceConfig
+}
+
+type GroupMemberStatus struct {
+	group net.IP
+	// localMembers is a set for the local Pod's interface name which has joined in the multicast group.
+	localMembers sets.String
+	lastProbe    time.Time
+	mutex        sync.RWMutex
+}
+
+// eventHandler process the multicast Group membership report or leave messages.
+func (c *Controller) eventHandler(stopCh <-chan struct{}) {
+	for {
+		select {
+		case e := <-c.groupEventCh:
+			c.addOrUpdateGroupEvent(e)
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+// addGroupMemberStatus adds the new group into groupCache.
+func (c *Controller) addGroupMemberStatus(e *mcastGroupEvent) {
+	status := &GroupMemberStatus{
+		group:        e.group,
+		lastProbe:    e.time,
+		localMembers: sets.NewString(e.iface.InterfaceName),
+	}
+	c.groupCache.Add(status)
+	c.queue.Add(e.group.String())
+	klog.InfoS("Added new multicast group to cache", "group", e.group, "interface", e.iface.InterfaceName)
+	return
+}
+
+// updateGroupMemberStatus updates the group status in groupCache. If a "join" message is sent from an existing member,
+// only updates the lastProbe time. If a "join" message is sent from an "unknown" member, updates the lastProbe time and
+// adds the new member into the group's local member set. If a "leave" message is sent from an existing member, remove it from the
+// group's local member set, and if the member is the last one in local cache, a query message on the group is sent out
+// to check if there are still local members in the group.
+func (c *Controller) updateGroupMemberStatus(obj interface{}, e *mcastGroupEvent) {
+	status := obj.(*GroupMemberStatus)
+	newStatus := &GroupMemberStatus{
+		group:        status.group,
+		localMembers: status.localMembers,
```

Comment:
Again, I'm pretty sure this will cause problems to the storage, its index in particular as I have commented in https://github.com/antrea-io/antrea/pull/2652#discussion_r750942206.
Please add a unit test of `groupLeave` and try to call `getGroupMemberStatusesByPod` with the removed pod, you will see the problem.

Reply:
sure, would change and add a unit test